From da011a985fce208c3ad4eeb13d635323b1cbfc8d Mon Sep 17 00:00:00 2001
From: Doug Goldstein <cardoe@cardoe.com>
Date: Tue, 13 Jan 2026 11:49:05 -0600
Subject: [PATCH 5/8] feat: skip invalid mac addr interfaces in redfish inspect
 add speed

Refactor the redfish interfaces inspection by moving it to its own
function, but preserving the behavior of not having an interfaces key in
the inspection data when it is empty. If the MAC is invalid for the
interface, skip it instead of adding it. Added the speed_mbps field to
match agent inspection.

Change-Id: I60cdf6e22b4ae4773e4497f0e0f10d795ef2cb6c
Signed-off-by: Doug Goldstein <cardoe@cardoe.com>
---
 ironic/drivers/modules/redfish/inspect.py     | 35 +++++++++++++++----
 .../drivers/modules/redfish/test_inspect.py   |  8 +++--
 2 files changed, 35 insertions(+), 8 deletions(-)

diff --git a/ironic/drivers/modules/redfish/inspect.py b/ironic/drivers/modules/redfish/inspect.py
index 69f0f0c09..9786c9f46 100644
--- a/ironic/drivers/modules/redfish/inspect.py
+++ b/ironic/drivers/modules/redfish/inspect.py
@@ -14,6 +14,7 @@ Redfish Inspect Interface
 """
 
 from oslo_log import log
+from oslo_utils import netutils
 from oslo_utils import units
 import sushy
 
@@ -141,12 +142,7 @@ class RedfishInspect(base.InspectInterface):
 
             inventory['disks'] = disks
 
-        if system.ethernet_interfaces and system.ethernet_interfaces.summary:
-            inventory['interfaces'] = []
-            for eth in system.ethernet_interfaces.get_members():
-                iface = {'mac_address': eth.mac_address,
-                         'name': eth.identity}
-                inventory['interfaces'].append(iface)
+        inventory['interfaces'] = self._get_interface_info(task, system)
 
         system_vendor = {}
         if system.model:
@@ -297,6 +293,33 @@ class RedfishInspect(base.InspectInterface):
         """
         return None
 
+    def _get_interface_info(self, task, system):
+        """Extract ethernet interface info."""
+
+        ret = []
+        if not system.ethernet_interfaces:
+            return ret
+
+        for eth in system.ethernet_interfaces.get_members():
+            if not netutils.is_valid_mac(eth.mac_address):
+                LOG.warning(_("Ignoring NIC address '%(address)s' for "
+                              "interface %(inf)s on node %(node)s because it "
+                              "is not a valid MAC"),
+                            {'address': eth.mac_address,
+                             'inf': eth.identity,
+                             'node': task.node.uuid})
+                continue
+            intf = {
+                'mac_address': eth.mac_address,
+                'name': eth.identity
+            }
+            try:
+                intf['speed_mbps'] = int(eth.speed_mbps)
+            except Exception:
+                pass
+            ret.append(intf)
+        return ret
+
     def _get_processor_info(self, task, system):
         # NOTE(JayF): Checking truthiness here is better than checking for None
         #             because if we have an empty list, we'll raise a
diff --git a/ironic/tests/unit/drivers/modules/redfish/test_inspect.py b/ironic/tests/unit/drivers/modules/redfish/test_inspect.py
index c81fb3934..4b8d08217 100644
--- a/ironic/tests/unit/drivers/modules/redfish/test_inspect.py
+++ b/ironic/tests/unit/drivers/modules/redfish/test_inspect.py
@@ -108,6 +108,7 @@ class RedfishInspectTestCase(db_base.DbTestCase):
             spec=sushy.resources.system.ethernet_interface.EthernetInterface)
         eth_interface_mock1.identity = 'NIC.Integrated.1-1'
         eth_interface_mock1.mac_address = '00:11:22:33:44:55'
+        eth_interface_mock1.speed_mbps = 25000
         eth_interface_mock1.status.state = sushy.STATE_ENABLED
         eth_interface_mock1.status.health = sushy.HEALTH_OK
 
@@ -115,6 +116,7 @@ class RedfishInspectTestCase(db_base.DbTestCase):
             spec=sushy.resources.system.ethernet_interface.EthernetInterface)
         eth_interface_mock2.identity = 'NIC.Integrated.2-1'
         eth_interface_mock2.mac_address = '66:77:88:99:AA:BB'
+        eth_interface_mock2.speed_mbps = 25000
         eth_interface_mock2.status.state = sushy.STATE_DISABLED
         eth_interface_mock2.status.health = sushy.HEALTH_OK
 
@@ -203,9 +205,11 @@ class RedfishInspectTestCase(db_base.DbTestCase):
                          system_vendor['system_uuid'])
 
         expected_interfaces = [{'mac_address': '00:11:22:33:44:55',
-                                'name': 'NIC.Integrated.1-1'},
+                                'name': 'NIC.Integrated.1-1',
+                                'speed_mbps': 25000},
                                {'mac_address': '66:77:88:99:AA:BB',
-                                'name': 'NIC.Integrated.2-1'}]
+                                'name': 'NIC.Integrated.2-1',
+                                'speed_mbps': 25000}]
         self.assertEqual(expected_interfaces,
                          inventory['inventory']['interfaces'])
 
-- 
2.50.1 (Apple Git-155)
